File: android/src/main/java/expo/modules/speechtotext/SpeechToTextTranscriptAccumulator.kt
package expo.modules.speechtotext

internal class SpeechToTextTranscriptAccumulator {
  private val committedSegments = mutableListOf<String>()
  private var pendingSegment: String? = null

  fun reset() {
    committedSegments.clear()
    pendingSegment = null
  }

  fun updatePending(segment: String) {
    pendingSegment = normalize(segment)
  }

  fun commitPending(): Boolean {
    val nextSegment = pendingSegment
    pendingSegment = null
    if (nextSegment == null) {
      return false
    }

    committedSegments += nextSegment
    return true
  }

  fun preview(): String? {
    val segments = pendingSegment?.let { committedSegments + it } ?: committedSegments
    return segments.joinToString(" ").trim().ifEmpty { null }
  }

  fun finalized(): String? = committedSegments.joinToString(" ").trim().ifEmpty { null }

  private fun normalize(segment: String): String? = segment.trim().ifEmpty { null }
}

File: ios/SpeechToTextTranscriptAccumulator.swift
import Foundation

struct SpeechToTextTranscriptAccumulator {
  private var committedSegments: [String] = []
  private var pendingSegment: String?

  mutating func reset() {
    committedSegments = []
    pendingSegment = nil
  }

  mutating func discardPending() {
    pendingSegment = nil
  }

  mutating func updatePending(_ segment: String) {
    let normalized = segment.trimmingCharacters(in: .whitespacesAndNewlines)
    pendingSegment = normalized.isEmpty ? nil : normalized
  }

  @discardableResult
  mutating func commitPending() -> Bool {
    guard let pendingSegment else {
      return false
    }

    committedSegments.append(pendingSegment)
    self.pendingSegment = nil
    return true
  }

  var preview: String? {
    normalizedTranscript(
      segments: pendingSegment.map { committedSegments + [$0] } ?? committedSegments
    )
  }

  var finalized: String? {
    normalizedTranscript(segments: committedSegments)
  }

  private func normalizedTranscript(segments: [String]) -> String? {
    let transcript = segments
      .joined(separator: " ")
      .trimmingCharacters(in: .whitespacesAndNewlines)
    return transcript.isEmpty ? nil : transcript
  }
}
